bpf/lib/nodeport.h

```diff
@@ -708,51 +711,78 @@ static __always_inline bool nodeport_uses_dsr4(const struct ipv4_ct_tuple *tuple
 }
 
 static __always_inline bool nodeport_nat_ipv4_needed(struct __ctx_buff *ctx,
-						     __be32 addr, int dir)
+						     __be32 addr, int dir,
+						     bool *from_endpoint __maybe_unused,
+						     const bool encap __maybe_unused)
 {
 	void *data, *data_end;
 	struct iphdr *ip4;
 
 	if (!revalidate_data(ctx, &data, &data_end, &ip4))
 		return false;
-#ifdef ENABLE_DSR_HYBRID
-	if (nodeport_uses_dsr(ip4->protocol))
-		return false;
-#endif /* ENABLE_DSR_HYBRID */
 	/* Basic minimum is to only NAT when there is a potential of
 	 * overlapping tuples, e.g. applications in hostns reusing
 	 * source IPs we SNAT in node-port.
 	 */
-	if (dir == NAT_DIR_EGRESS)
-		return ip4->saddr == addr;
-	else
+	if (dir == NAT_DIR_EGRESS) {
+		if (ip4->saddr == addr) {
+			return true;
+		}
+#ifdef ENABLE_MASQUERADE
+		if (encap)
+			return false;
+
+		// TODO(brb) use skb mark to determine if src is a local endpoint
+		if (__lookup_ip4_endpoint(ip4->saddr) != NULL) {
```

Comment:
Hmm, good point. I think we should only MASQ outgoing connections from endpoints (changed). Do you know a case where a packet egressing to the world would have `cilium_host` IP as src?

Reply:
I don't know any such case. `cilium_host` IP should only be chosen for source when traffic from the host is routed towards the `cilium_host` device, which is typically only the CIDR prefixes that Cilium manages.

The kind of case I'd be concerned about is something like if we have a virtual IP / DNAT type feature and set up routes for those destinations. However we don't route services to `cilium_host` anyway AFAIK, we will only do such service translation from the host-reachable services socket layer hooks. So that should not occur.